cpp/src/experimental/induced_subgraph.cu

```diff
@@ -0,0 +1,280 @@
+        thrust::tie(indices, weights, local_degree) =
+          matrix_partition.get_local_edges(major_offset);
+        if (weights != nullptr) {
+          auto triplet_first = thrust::make_zip_iterator(thrust::make_tuple(
+            thrust::make_constant_iterator(subgraph_vertices[i]), indices, weights));
+          // FIXME: this is inefficient for high local degree vertices
+          thrust::copy_if(
+            thrust::seq,
+            triplet_first,
+            triplet_first + local_degree,
+            thrust::make_zip_iterator(thrust::make_tuple(edge_majors, edge_minors, edge_weights)) +
+              subgraph_vertex_output_offsets[i],
+            [vertex_first = subgraph_vertices + subgraph_offsets[subgraph_idx],
+             vertex_last =
+               subgraph_vertices + subgraph_offsets[subgraph_idx + 1]] __device__(auto t) {
+              return thrust::binary_search(
+                thrust::seq, vertex_first, vertex_last, thrust::get<1>(t));
+            });
+        } else {
+          auto pair_first = thrust::make_zip_iterator(
+            thrust::make_tuple(thrust::make_constant_iterator(subgraph_vertices[i]), indices));
+          // FIXME: this is inefficient for high local degree vertices
+          thrust::copy_if(thrust::seq,
+                          pair_first,
+                          pair_first + local_degree,
+                          thrust::make_zip_iterator(thrust::make_tuple(edge_majors, edge_minors)) +
+                            subgraph_vertex_output_offsets[i],
+                          [vertex_first = subgraph_vertices + subgraph_offsets[subgraph_idx],
+                           vertex_last  = subgraph_vertices +
+                                         subgraph_offsets[subgraph_idx + 1]] __device__(auto t) {
+                            return thrust::binary_search(
+                              thrust::seq, vertex_first, vertex_last, thrust::get<1>(t));
+                          });
+        }
+      });
+
+    rmm::device_uvector<size_t> subgraph_edge_offsets(num_subgraphs + 1, handle.get_stream());
+    thrust::gather(rmm::exec_policy(handle.get_stream())->on(handle.get_stream()),
+                   subgraph_offsets,
+                   subgraph_offsets + (num_subgraphs + 1),
+                   subgraph_vertex_output_offsets.begin(),
+                   subgraph_edge_offsets.begin());
+
+    CUDA_TRY(
+      cudaStreamSynchronize(handle.get_stream()));  // subgraph_vertex_output_offsets will become
+                                                    // out-of-scope once this function returns
+
+    return std::make_tuple(std::move(edge_majors),
+                           std::move(edge_minors),
+                           std::move(edge_weights),
+                           std::move(subgraph_edge_offsets));
+  }
+}
+
+// explicit instantiation
+
+template std::tuple<rmm::device_uvector<int32_t>,
```

Comment:
Yes, so what type combinations do you need?; we need to include all the types we actually use (for the obvious reason) but better avoid instantiating for any types we don't use (as this increases compile-time and binary size).

for (vertex_t, edge_t, weight_t) triplets, (int32_t, int32_t, float), (int32_t, int64_t, float), (int64_t, int64_t, float) must be instantiated I guess, and are we actually using double weights?

Reply:
Yes, I have instantiated these in Egonet: https://github.com/afender/cugraph/blob/7844fa4c35500fb85d9f38a9b2f74d640684fc9b/cpp/src/community/egonet.cu#L128
For the double weights, it is a good discussion. I don't think it is motivated by this algo but since the graph and other algos accept it, we should probably instantiate it here as well.